### ISSUE TITLE
feature: Run lint tools within a Docker container

### ISSUE THREAD

**Issue by WillDaSilva** [meltano/meltano#7461]
### Feature scope

Other

### Description

Running `pre-commit` requires it be installed using Python, which is a reasonable requirement for Python developers who already need a functional Python installation anyway, but is an undue burden e.g. on contributors updating docs. In the past we've seen issues with `pyenv` (e.g. `zlib` missing on macOS), `pipx` not working properly (thereby preventing `pipx run pre-commit`, and much more.

To simplify running our lint checks, we could offer a containerized version using a script `scripts/lint.sh`, which when run essentially does `docker run ...` to run `pre-commit run --all-files` with the repository mounted in (so that the checks can apply auto-fixes), running as the calling user and group so that the auto-fixed files aren't owned by root.

**Comment by aaronsteers** [comment]
I was able to get my issue resolved by uninstalling and reinstalling Pyenv and Python. (Less work than it sounds like.)

https://meltano.slack.com/archives/C03GKHWS0HM/p1680123535468659?thread_ts=1680115248.635779&cid=C03GKHWS0HM

Something like this:

```
brew upgrade pyenv
pyenv install 3.9      # Installed 3.9.16
pyenv remove 3.8.10    # Just to be sure
pipx reinstall poetry  # Because it was looking for the old interpreter
```

**Comment by aaronsteers** [comment]
I think we can deprioritize this or close if not needed. Before going forward, I'd want to make sure we don't run a chance of the same/similar issue occurring the docker layer, since now I think we have a path to recommend a straightforward fix for Python.

For example, on the docker side, I recently moved to Rancher instead of Docker Desktop, and I know there's a lot of different options out there these days. Before migrating over to container-based, we'd want to consider if there's a chance of interop challenges. Not saying personally that I think there will be, but just that every layer introduces a chance that someone will struggle with interop based on one issue or another. 🙃

**Comment by WillDaSilva** [comment]
Closing as not planned. I have a working solution for everything except for the eslint check, which requires a more modern version of NodeJS than what is available by default on `python:3.10-slim`. Getting that working using nvm is proving to be a bit of a pain...

Besides that, if no one is using this lint script it will definitely fall out of date, so without someone else asking for it there's little reason to add it.

For posterity, here's the mostly working approach:

`scripts/lint.sh`:

```sh
#!/usr/bin/env bash

SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
mkdir -p "${HOME}/.cache/pre-commit"
docker run --rm -it --entrypoint bash \
    -u "$(id -u):$(id -g)" \
    -v "${SCRIPT_DIR}/..:/project" \
    -v "${HOME}/.cache/pre-commit:/.cache/pre-commit" \
    "$(docker build -q ${SCRIPT_DIR}/docker/lint)"
```

`scripts/docker/lint/Dockerfile`:

```dockerfile
FROM python:3.10-slim

RUN apt-get update && \
    apt-get install -y bash build-essential curl git golang libbz2-dev libffi-dev libgdbm-dev liblzma-dev libncursesw5-dev libreadline-dev libsqlite3-dev libssl-dev python3-distutils tk-dev uuid-dev zlib1g-dev && \
    rm -rf /var/cache/apt/archives /var/lib/apt/lists/*

RUN mkdir -p '/.cache/pre-commit'

ENV PRE_COMMIT_HOME='/.cache/pre-commit'

ARG PRE_COMMIT_VERSION=3.2.1

RUN pip install "pre-commit==${PRE_COMMIT_VERSION}"

WORKDIR /project

CMD ["pre-commit", "run", "--all-files"]
```
